File: LaRC/CER_SSF_NPP_FM5_VIIRS_Edition2A_200203_2022013123_nc.m
%
%  This example code illustrates how to access and visualize
% LaRC CERES SSF NPP L2 netCDF-4/HDF5 file in MATLAB. 
%
%  If you have any questions, suggestions, comments  on this example, 
% pleaseuse the HDF-EOS Forum (http://hdfeos.org/forums). 
%
%  If you would like to see an  example of any other NASA
% HDF/HDF-EOS data product that is not listed in the
% HDF-EOS Comprehensive Examples page (http://hdfeos.org/zoo), 
% feel free to contact us at eoshelp@hdfgroup.org or post it at the
% HDF-EOS Forum (http://hdfeos.org/forums).
%
% Usage:save this script and run (without .m at the end)
%
%  $matlab -nosplash -nodesktop -r CER_SSF_NPP_FM5_VIIRS_Edition2A_200203_2022013123_nc
%
% Tested under: MATLAB R2021a
% Last updated: 2022-10-18

% Open the netCDF-4/HDF5 file.
FILE_NAME='CER_SSF_NPP-FM5-VIIRS_Edition2A_200203.2022013123.nc';

% List file content.
% ncdisp(FILE_NAME)

ncid = netcdf.open(FILE_NAME, 'nowrite');

% Read data from a data field.
DATAFIELD_NAME='/TOA_and_Surface_Fluxes/model_a_clearsky_surface_longwave_downward_flux';

data = ncread(FILE_NAME, DATAFIELD_NAME);

% Read units.
units = ncreadatt(FILE_NAME, DATAFIELD_NAME, "units");

% Read long_name.
long_name = ncreadatt(FILE_NAME, DATAFIELD_NAME, "long_name");


% Read latitude data.
DATAFIELD_NAME='/Time_and_Position/instrument_fov_latitude';
lat = ncread(FILE_NAME, DATAFIELD_NAME);
lat = double(lat);

% Read longitude data.
DATAFIELD_NAME='/Time_and_Position/instrument_fov_longitude';
lon = ncread(FILE_NAME, DATAFIELD_NAME);
lon = double(lon);

% Close the file.
netcdf.close(ncid);

% Create the graphics figure.
f=figure('Name', FILE_NAME, ...
	 'Renderer', 'zbuffer', ...
	 'Position', [0,0,800,600], ...
	 'visible', 'off');

% Plot the data using surfm and axesm.
axesm('MapProjection','eqdcylin','Frame','on','Grid','on', ...
      'MeridianLabel','on','ParallelLabel','on', 'MLabelParallel','south', ...
      'FontSize', 8)

scatterm(lat, lon, 1, data);
colormap('Jet');
min_data=floor(min(min(data)));
max_data=ceil(max(max(data)));
caxis([min_data max_data]); 

coast = load('coastlines.mat');
plotm(coast.coastlat,coast.coastlon,'k')

% Change the value if you want to have more than 10 tick marks.
h = colorbar();
set (get(h, 'title'), 'string', units)
tightmap;

title({FILE_NAME;long_name}, 'Interpreter', 'None', ...
      'FontSize',10,'FontWeight','bold');


saveas(f, [FILE_NAME  '.m.png']);
exit

File: LaRC/CER_SSF_NPP-FM5-VIIRS_Edition2A_200203.2022013123.nc.py
"""
This example code illustrates how to access and visualize a LaRC CERES SSF
NPP L2 netCDF-4/HDF5 file in Python.

If you have any questions, suggestions, or comments on this example, please use
the HDF-EOS Forum (http://hdfeos.org/forums).  If you would like to see an
example of any other NASA HDF/HDF-EOS data product that is not listed in the
HDF-EOS Comprehensive Examples page (http://hdfeos.org/zoo), feel free to
contact us at eoshelp@hdfgroup.org or post it at the HDF-EOS Forum
(http://hdfeos.org/forums).

Usage:  save this script and run

    $python CER_SSF_NPP-FM5-VIIRS_Edition2A_200203.2022013123.nc.py

The netCDF-4/HDF5 file must either be in your current working directory.

Tested under: Python 3.9.12 :: Miniconda
Last updated: 2022-10-10

"""
import os
import h5py

import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt

from mpl_toolkits.basemap import Basemap

FILE_NAME = 'CER_SSF_NPP-FM5-VIIRS_Edition2A_200203.2022013123.nc'
with h5py.File(FILE_NAME, mode='r') as f:

    latvar = f['/Time_and_Position/instrument_fov_latitude']
    latitude = latvar[:]
    
    lonvar = f['/Time_and_Position/instrument_fov_longitude']
    longitude = lonvar[:]
    
    dset_name = '/TOA_and_Surface_Fluxes/model_a_clearsky_surface_longwave_downward_flux'
    datavar = f[dset_name]
    data = np.float32(datavar[:])
    units = datavar.attrs['units']
    units = units.decode('ascii', 'replace')
    
    long_name = datavar.attrs['long_name']
    long_name = long_name.decode('ascii', 'replace')
        
    _FillValue = datavar.attrs['_FillValue']
    
    # Handle fill values.
    data[data == _FillValue] = np.nan
    data = np.ma.masked_where(np.isnan(data), data)

    m = Basemap(projection='cyl', resolution='l',
                llcrnrlat=-90, urcrnrlat=90,
                llcrnrlon=-180, urcrnrlon=180)
    m.scatter(longitude, latitude, c=data, s=1, cmap=plt.cm.jet,
              edgecolors=None, linewidth=0)
    m.drawcoastlines(linewidth=0.5)
    m.drawparallels(np.arange(-90, 91, 45))
    m.drawmeridians(np.arange(-180, 180, 45), labels=[True,False,False,True])
    cb = m.colorbar()
    cb.set_label(units)
    basename = os.path.basename(FILE_NAME)
    plt.title('{0}\n{1}'.format(basename, long_name), fontsize=8)
    pngfile = "{0}.py.png".format(basename)
    fig = plt.gcf()
    fig.savefig(pngfile)


# Reference
#  [1] https://cmr.earthdata.nasa.gov/search/concepts/C2246001739-LARC_ASDC.html

